server/commands/list_objects.go

```diff
@@ -0,0 +1,232 @@
+	uniqueObjects, err := q.datastore.ReadUniqueObjects(ctx, storage.ReadRelationshipTuplesFilter{
+		StoreID:            storeID,
+		OptionalObjectType: targetObjectType,
+	})
+	if err != nil {
+		return nil, serverErrors.NewInternalError("", err)
+	}
+
+	if ctxTuples != nil {
+		// Include the objects in the contextual tuples in the result
+		uniqueSet := make(map[string]bool, len(uniqueObjects))
+		for _, o := range uniqueObjects {
+			_, found := uniqueSet[o]
+			if !found {
+				uniqueSet[o] = true
+			}
+		}
+
+		for _, ctxTuple := range ctxTuples.TupleKeys {
+			_, found := uniqueSet[ctxTuple.Object]
+			if !found {
+				uniqueSet[ctxTuple.Object] = true
+				uniqueObjects = append(uniqueObjects, ctxTuple.Object)
+			}
+		}
+	}
+	return uniqueObjects, nil
+}
+
+func (q *ListObjectsQuery) validateInput(ctx context.Context, storeID string, targetObjectType string, authModelID string, relation string) error {
+	definition, err := q.datastore.ReadTypeDefinition(ctx, storeID, authModelID, targetObjectType)
+	if err != nil {
+		if errors.Is(err, storage.ErrNotFound) {
+			return serverErrors.TypeNotFound(targetObjectType)
+		}
+		return err
+	}
+	_, ok := definition.Relations[relation]
+	if !ok {
+		return serverErrors.UnknownRelationWhenListingObjects(relation, targetObjectType)
+	}
+	return nil
+}
+
+func (q *ListObjectsQuery) performChecks(timeoutCtx context.Context, storeID string, authModelID string, objectType string, relation string, user string, ctxTuples *openfgapb.ContextualTupleKeys, resultsChan chan<- string, errChan chan<- error, resolvedChan chan<- struct{}) {
+	g := new(errgroup.Group)
+	g.SetLimit(MaximumConcurrentChecks)
+	var objectsFound uint32
+
+	uniqueObjects, err := q.getUniqueObjects(timeoutCtx, storeID, objectType, ctxTuples)
+	if err != nil {
+		errChan <- err
+		close(errChan)
+		return
+	}
+
+	// iterate over all object IDs in the store and check if the user has relation with each
+	for _, object := range uniqueObjects {
+		object := object
+		if atomic.LoadUint32(&objectsFound) >= q.listObjectsMaxResults {
+			break
+		}
+
+		checkFunction := func() error {
+			_, objectID := tuple.SplitObject(object)
+			query := NewCheckQuery(q.datastore, q.tracer, q.meter, q.logger, q.resolveNodeLimit)
+
+			resp, err := query.Execute(timeoutCtx, &openfgapb.CheckRequest{
+				StoreId:              storeID,
+				AuthorizationModelId: authModelID,
+				TupleKey: &openfgapb.TupleKey{
+					Object:   object,
+					Relation: relation,
+					User:     user,
+				},
+				ContextualTuples: ctxTuples,
+			})
+			if err != nil {
+				return err
+			}
+
+			if resp.Allowed && atomic.LoadUint32(&objectsFound) < q.listObjectsMaxResults {
+				resultsChan <- objectID
+				atomic.AddUint32(&objectsFound, 1)
+			}
+
+			return nil
+		}
+
+		g.Go(checkFunction)
+	}
+
+	err = g.Wait()
+	if err != nil {
+		// we don't want to abort everything if one of the checks failed.
+		q.logger.Warn("ListObjects errored: ", logger.Error(err))
```

Comment:
If that is the case you should use the error channel within `checkFunction` otherwise in the first error it get's it will exit.

Reply:
https://github.com/openfga/openfga/pull/145/files#diff-7f35d19e9a3e7af70b152b10bcb55e9906ba5fbece03b60a5a03dbc2d1d9ec1bR209

Should return `nil` but write to the error channel.

Reply:
No, if I write to the error channel it will have the same effect because of the `case error <- errChan`... 

I'll return nil and just log the error